File: atlas-perception/Thermal/ThermalSensor.ts
import { seededRange } from "../../atlas-kernel/utils/deterministic";

export interface ThermalFrame {
  width: number;
  height: number;
  temperatureData: Float32Array;
  timestamp: number;
}

export class ThermalSensor {
  private config: { width: number; height: number };
  private isRunning: boolean;
  private frameCallback?: (frame: ThermalFrame) => void;
  private frameIndex = 0;

  constructor() {
    this.config = { width: 320, height: 240 };
    this.isRunning = false;
  }

  async start(): Promise<void> {
    this.isRunning = true;
  }

  async stop(): Promise<void> {
    this.isRunning = false;
  }

  captureFrame(): ThermalFrame {
    const { width, height } = this.config;
    const numPixels = width * height;
    const tempData = new Float32Array(numPixels);
    const t = this.frameIndex * 0.1;
    const horizonY = Math.floor(height * 0.5);

    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const i = y * width + x;
        const seed = this.frameIndex * numPixels + i;
        let temp: number;

        if (y < horizonY - 3) {
          const grad = y / (horizonY - 3);
          temp = -5 + grad * 10 + seededRange(seed, -1.5, 1.5);
          if (seededRange(seed + 1000, 0, 1) > 0.97) {
            temp += 5 + seededRange(seed + 2000, 0, 8);
          }
        } else if (y > horizonY + 3) {
          const grad = (y - horizonY) / (height - horizonY);
          temp = 22 + grad * 8 + seededRange(seed, -1, 1);
          if (seededRange(seed + 3000, 0, 1) > 0.96) {
            temp += 10 + seededRange(seed + 4000, 0, 15);
          }
        } else {
          temp = 15 + seededRange(seed, -3, 3);
        }

        const hx = Math.floor(width * 0.6);
        const hy = Math.floor(height * 0.25);
        if (Math.abs(x - hx) < 8 && Math.abs(y - hy) < 8) {
          const dist = Math.hypot(x - hx, y - hy);
          if (dist < 8) temp += 25 * (1 - dist / 8);
        }

        const ex = Math.floor(width * 0.3);
        const ey = Math.floor(height * 0.7);
        if (Math.abs(x - ex) < 5 && Math.abs(y - ey) < 12) {
          const dist = Math.hypot((x - ex) / 5, (y - ey) / 12);
          if (dist < 1) temp += 18 * (1 - dist);
        }

        tempData[i] = temp;
      }
    }

    this.frameIndex += 1;
    const frame = {
      width,
      height,
      temperatureData: tempData,
      timestamp: Date.now(),
    };
    this.frameCallback?.(frame);
    return frame;
  }

  setFrameCallback(callback: (frame: ThermalFrame) => void): void {
    this.frameCallback = callback;
  }
}

File: atlas-perception-py/atlas_perception/thermal/thermal_sensor.py
import math
import time
from typing import Callable, Optional

import numpy as np

from ..deterministic import seeded_range
from ..types import ThermalFrame


class ThermalSensor:
    def __init__(self):
        self._config = {"width": 320, "height": 240}
        self._is_running = False
        self._frame_callback: Optional[Callable[[ThermalFrame], None]] = None
        self._frame_index = 0

    async def start(self) -> None:
        self._is_running = True

    async def stop(self) -> None:
        self._is_running = False

    def capture_frame(self) -> ThermalFrame:
        width = self._config["width"]
        height = self._config["height"]
        num_pixels = width * height
        temp_data = np.zeros(num_pixels, dtype=np.float32)
        t = self._frame_index * 0.1
        horizon_y = int(height * 0.5)

        for y in range(height):
            for x in range(width):
                i = y * width + x
                seed = self._frame_index * num_pixels + i
                if y < horizon_y - 3:
                    grad = y / (horizon_y - 3) if horizon_y > 3 else 0
                    temp = -5 + grad * 10 + seeded_range(seed, -1.5, 1.5)
                    if seeded_range(seed + 1000, 0, 1) > 0.97:
                        temp += 5 + seeded_range(seed + 2000, 0, 8)
                elif y > horizon_y + 3:
                    grad = (y - horizon_y) / (height - horizon_y) if height > horizon_y else 0
                    temp = 22 + grad * 8 + seeded_range(seed, -1, 1)
                    if seeded_range(seed + 3000, 0, 1) > 0.96:
                        temp += 10 + seeded_range(seed + 4000, 0, 15)
                else:
                    temp = 15 + seeded_range(seed, -3, 3)
                hx = int(width * 0.6)
                hy = int(height * 0.25)
                dx_h = x - hx
                dy_h = y - hy
                if abs(dx_h) < 8 and abs(dy_h) < 8:
                    dist = math.hypot(dx_h, dy_h)
                    if dist < 8:
                        temp += 25 * (1 - dist / 8)
                ex = int(width * 0.3)
                ey = int(height * 0.7)
                dx_e = x - ex
                dy_e = y - ey
                if abs(dx_e) < 5 and abs(dy_e) < 12:
                    dist = math.hypot(dx_e / 5, dy_e / 12)
                    if dist < 1:
                        temp += 18 * (1 - dist)
                temp_data[i] = temp

        self._frame_index += 1
        frame = ThermalFrame(
            width=width,
            height=height,
            temperatureData=temp_data,
            timestamp=time.time(),
        )
        if self._frame_callback:
            self._frame_callback(frame)
        return frame

    def set_frame_callback(self, callback: Callable[[ThermalFrame], None]) -> None:
        self._frame_callback = callback
